pyiron_atomistics/atomistics/structure/atoms.py

```diff
@@ -2275,29 +2275,45 @@ def set_constraint(self, constraint=None):
                 else:
                     self.selective_dynamics[atom_ind] = [True, True, True]
 
-    def apply_strain(self, epsilon, return_box=False):
+    def apply_strain(self, epsilon, return_box=False, mode='linear'):
         """
-        Apply a given strain on the structure
+        Apply a given strain on the structure. It applies the matrix `F` in the manner:
+
+        ```
+            new_cell = F @ current_cell
+        ```
 
         Args:
-            epsilon (float/list/ndarray): epsilon matrix. If a single number is set, the same strain
-                                          is applied in each direction. If a 3-dim vector is set, it
-                                          will be multiplied by a unit matrix.
+            epsilon (float/list/ndarray): epsilon matrix. If a single number is set, the same
+                strain is applied in each direction. If a 3-dim vector is set, it will be
+                multiplied by a unit matrix.
             return_box (bool): whether to return a box. If set to True, only the returned box will
-                               have the desired strain and the original box will stay unchanged.
+                have the desired strain and the original box will stay unchanged.
+            mode (str): `linear` or `lagrangian`. If `linear`, `F` is equal to the epsilon value.
+                If `lagrangian`, epsilon is given by `(F^T * F - 1) / 2`. It raises an error if
+                the strain is not symmetric (if the shear components are given).
         """
+        if mode not in ['lagrangian', 'linear']:
+            raise ValueError('mode must be `linear` or `lagrangian`')
         epsilon = np.array([epsilon]).flatten()
         if len(epsilon) == 3 or len(epsilon) == 1:
             epsilon = epsilon*np.eye(3)
         epsilon = epsilon.reshape(3, 3)
         if epsilon.min() < -1.0:
             raise ValueError("Strain value too negative")
+        if mode == 'lagrangian' and not np.allclose(epsilon, epsilon.T):
+            raise ValueError("Strain must be symmetric if `mode = 'lagrangian'`")
         if return_box:
             structure_copy = self.copy()
         else:
             structure_copy = self
         cell = structure_copy.cell.copy()
-        cell = np.matmul(epsilon + np.eye(3), cell)
+        if mode == 'linear':
+            F = epsilon + np.eye(3)
+        else:
+            E, V = np.linalg.eigh(2*epsilon+np.eye(3))
```

Comment:
Minor nit: I know that `np.einsum` is an efficient way to do tensor/matrix operations but it is hard to understand what you end up with on first glance. Perhaps add a comment explaining whats going on would be nice.

Reply:
The problem for me is that while many people told me that they have trouble with `einsum`, no one made me understand what they find difficult ... Is it the Einstein notation itself? Or do you prefer simply a completely different notation like this (to be honest this one below is super difficult to read for me)?

```python
F = np.sum(V[:, np.newaxis, :] * np.sqrt(E)[np.newaxis, np.newaxis, :] * V.T[np.newaxis, :, :], axis=1)
```

By the way, mostly the reason I use `einsum` is because it's easy to understand which axes are multiplies with which axes & which axes are summed, and less because of its efficiency...

Reply:
> 
> 
> The problem for me is that while many people told me that they have trouble with `einsum`, no one made me understand what they find difficult ... Is it the Einstein notation itself? Or do you prefer simply a completely different notation like this (to be honest this one below is super difficult to read for me)?
> 
> ```python
> F = np.sum(V[:, np.newaxis, :] * np.sqrt(E)[np.newaxis, np.newaxis, :] * V.T[np.newaxis, :, :], axis=1)
> ```

From this line, I can now tell that you reshape the eigenvalue and eigenvectors to perform two element-wise multiplication operations and then sum over one axis. For me personally, the logic is more apparent in this example because I am not too familiar with the `einsum` notation (which is my fault). I agree with you that using `einsum` is cleaner and more `numpy` like. I just thought having a comment explaining what you do would help people like me.